argilla-server/src/argilla_server/bulk/records_bulk.py

```diff
@@ -192,7 +192,7 @@ async def upsert_records_bulk(self, dataset: Dataset, bulk_upsert: RecordsBulkUp
         records = []
         async with self._db.begin_nested():
             for record_upsert in bulk_upsert.items:
-                record = found_records.get(record_upsert.external_id or record_upsert.id)
+                record = found_records.get(record_upsert.external_id) or found_records.get(record_upsert.id)
```

Comment:
Small possible performance improvement: if it's more common to use the `id` instead of the `external_id` we should put the one getting the `id` first.